File: week5/python/Leaf_SimilarTrees_872.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def leafSimilar(self, root1: TreeNode, root2: TreeNode) -> bool:
        
        leaf_sequence = list()
        def traverse(root: TreeNode, found: bool):
            if not root:
                return True
            
            if not root.left and not root.right:
                if found:
                    if leaf_sequence[0] != root.val:
                        return False
                    leaf_sequence[:] = leaf_sequence[1:]
                else:
                    leaf_sequence.append(root.val)
                return True
            
            return traverse(root.left, found) \
                    and traverse(root.right, found)
            
        traverse(root1, False)
        return traverse(root2, True)

File: week5/java/Leaf_SimilarTrees_872.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    
    private Queue<Integer> queue = new LinkedList<Integer>();
    
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        if (root1 == null || root2 == null) 
            return root1 == root2;
        
        traverse(root1, false);
        return traverse(root2, true) && queue.isEmpty();
    }
    
    private boolean traverse(TreeNode root, boolean isCheck) {
        if (root == null) return true;
        if (root.left == null && root.right == null) {
            if (isCheck) {
                return !queue.isEmpty() && queue.poll() == root.val;
            }
            queue.offer(root.val);
            return true;
        }
        return traverse(root.left, isCheck) && traverse(root.right, isCheck);
    }
}
